### PR TITLE
22803 - Pasting text or link in edit message pastes the text or link in main compose box

### PR DESCRIPTION
**Pull request by robertKozik**
### Details
it's follow up fix for https://github.com/Expensify/App/pull/21583.The copy&paste handler wasn't checking whether the incoming event fired from text input/area. It should be handled in a similar manner to the document-scoped keyPressListner

### Fixed Issues

$ https://github.com/Expensify/App/issues/22803
PROPOSAL: 

### Tests

1. Open the app
2. Open any report
3. Send any message
4. Copy any text
5. Edit message sent in step 3
6. Paste the text in edit message and observe that text is pasted in compose box

- [X] Verify that no errors appear in the JS console

### Offline tests

Same as test steps

### QA Steps

Same as test steps
- [X] Verify that no errors appear in the JS console

### PR Author Checklist

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I ran the tests on **all platforms** & verified they passed on:
    - [X] Android / native
    - [X] Android / Chrome
    - [X] iOS / native
    - [X] iOS / Safari
    - [X] MacOS / Chrome / Safari
    - [X] MacOS / Desktop
- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [X] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] I verified that if a function's arguments changed that all usages have also been updated correctly
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [X] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/61146594/e4e06575-0290-461f-9540-3fb30b9d4bd2

</details>

<details>
<summary>Mobile Web - Chrome</summary>

</details>

<details>
<summary>Mobile Web - Safari</summary>

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/61146594/767b5491-c9bc-4a22-8b98-bb97bdbf0f84

</details>

<details>
<summary>iOS</summary>

</details>

<details>
<summary>Android</summary>

</details>

### CODE REVIEW

**Comment by robertKozik** [comment]
I've record the videos for platforms which handle copy&pasting without focus. Let me know if I should record all others.

**Comment by Santhosh-Sellavel** [comment]
On it

**Comment by cristipaval** [comment]
> On it

Thanks for the urgency @Santhosh-Sellavel !

**Comment by Santhosh-Sellavel** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web & Desktop </summary>

https://github.com/Expensify/App/assets/85645967/cd37f6be-1d31-47ed-9dee-d8caba983f86

</details>

<details>
<summary>Mobile Web - Chrome & Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/85645967/52d79330-c06c-4a30-9774-decd0e0f61ad

</details>

**Review by Santhosh-Sellavel — APPROVED** [review]
LGTM, all yours @PauloGasparSv!

**Review by cristipaval — APPROVED** [review]

**Comment by francoisl** [comment]
Thanks for the quick fix everyone.

@cristipaval do you mind requesting a CP for this in #deployer please? We're trying to close to close the checklist ASAP today.

**Comment by francoisl** [comment]
Actually nevermind, I got this 😄

File: src/components/Composer/index.js [hunk 1]
```diff
@@ -306,7 +306,11 @@ class Composer extends React.Component {
      * @param {ClipboardEvent} event
      */
     handlePaste(event) {
-        if (!this.props.checkComposerVisibility() && !this.state.isFocused) {
+        if (!this.props.checkComposerVisibility()) {
+            return;
+        }
+
+        if (['INPUT', 'TEXTAREA'].includes(event.target.nodeName)) {
```
Comment (Santhosh-Sellavel) [inline src/components/Composer/index.js:313]: @robertKozik This is flawed in the root itself. ` this.props.checkComposerVisibility()` always fails for edit message items as we didn't pass `checkComposerVisibility`

Comment (Santhosh-Sellavel) [reply src/components/Composer/index.js:313]: This event is called `twice` one from each instance. We should perform the operation only when it's from the `textInput`

Comment (Santhosh-Sellavel) [reply src/components/Composer/index.js:313]: I came up with a quick solution here
```
if (!this.props.checkComposerVisibility() && !this.textInput.isFocused()) {
            return;
        }

        if (this.textInput.id !== event.target.id) {
            return;
        }
```

Comment (robertKozik) [reply src/components/Composer/index.js:313]: Yeah you're right. I'm also looking for the another fix for this one. Did you check if pasting to the edited comment works?

Comment (robertKozik) [reply src/components/Composer/index.js:313]: I'm not sure that paste event should be trigger only from focused input. The "parent" issue and PR https://github.com/Expensify/App/pull/21583 was introducing the ability to paste when input is not composed

Comment (Santhosh-Sellavel) [reply src/components/Composer/index.js:313]: No, it didn't work.

Comment (robertKozik) [reply src/components/Composer/index.js:313]: also, `checkComposerVisiblity` would always be true when no function is passed as the default prop always returns false.

Currently I'm experimenting with introducing textInput-scoped paste listener alongside the document-scoped one. When event is triggered from textInput/textArea document-scoped will do nothing, and textInput-scoped will do the hard work

Comment (Santhosh-Sellavel) [reply src/components/Composer/index.js:313]: Here is the snippet
        ```js
const isComposerVisible = this.props.checkComposerVisibility() 
        const isFocused = this.textInput.isFocused()
        if (!(isComposerVisible || isFocused)) {
            return;
        }

```

Comment (Santhosh-Sellavel) [reply src/components/Composer/index.js:313]: @robertKozik This works for me. Address me all cases.

cc: @francoisl @mountiny

Comment (francoisl) [reply src/components/Composer/index.js:313]: This snippet addresses both issues, but now if I edit an existing message and paste from the clipboard, the cursor remains at the beginning of the pasted message:

https://github.com/Expensify/App/assets/2229301/d6126975-0631-417e-aab6-6c47c80ee72a

I'd say let's revert this for now, and then work on a proper fix without having to rush.

Comment (Santhosh-Sellavel) [reply src/components/Composer/index.js:313]: Yes agreed lets do that first!

Comment (robertKozik) [reply src/components/Composer/index.js:313]: Great @Santhosh-Sellavel It works for me too

Comment (Santhosh-Sellavel) [reply src/components/Composer/index.js:313]: This [one](https://github.com/Expensify/App/pull/22817#discussion_r1263034541) comes out of nowhere.

Please revert this first and work on the other issue!

Comment (francoisl) [reply src/components/Composer/index.js:313]: Revert open here https://github.com/Expensify/App/pull/22845, feel free to review.

Comment (Santhosh-Sellavel) [reply src/components/Composer/index.js:313]: @robertKozik about the cursor [issue](https://github.com/Expensify/App/pull/22817/files#r1263034541)

The contributor [here](https://github.com/Expensify/App/issues/22833#issuecomment-1635174418) found the PR [#22443](https://github.com/Expensify/App/pull/22443) which causes it

**Pull request merged**